src/neorados/RADOSImpl.h

```diff
@@ -97,19 +100,39 @@ class Client {
   virtual int get_instance_id() const = 0;
 };
 
-class NeoClient : public Client {
+class NeoClient : public Client,
+		  public ceph::async::service_list_base_hook {
 public:
+
   NeoClient(std::unique_ptr<RADOS>&& rados)
     : Client(rados->ioctx, rados->cct, rados->monclient,
-             rados->objecter.get()),
+	     rados->objecter.get()),
+      svc(boost::asio::use_service<ceph::async::service<NeoClient>>(
+	  boost::asio::query(ioctx.get_executor(),
+			     boost::asio::execution::context))),
       rados(std::move(rados)) {
+    svc.add(*this);
+  }
+
+  ~NeoClient() {
+    svc.remove(*this);
   }
 
   int get_instance_id() const override {
     return rados->instance_id;
   }
 
+  void service_shutdown() {
+    rados->shutdown();
+    service_ref.reset();
+  }
+
+  // You have to set this after constructing the object!
+  std::shared_ptr<Client> service_ref;
```

Comment:
very strange for this to hold a reference to itself. if `service_shutdown()` is the only place that drops this ref, that means that _all_ short-lived neorados clients created over the duration of the process will be kept alive until shutdown

with your new `RADOS::consign()` function, all completion handlers should be holding a reference to this. isn't that sufficient to preserve its lifetime?

Reply:
it looks like all the other neorados tests pass without it and only `ceph_test_datalog` crashes without it, so I'll fix whatever's up with the test.